### DISCUSSION TITLE
Largo Saving stuck due to WebSocket connection failed

### DISCUSSION THREAD

**Discussion by cywhale** [biigle/core#563] [Q&A] [answered]
When performing Largo re-evaluation anotations of a volume, clicking Saving, the saving stuck and spinner keeps runnings.
No system logs of errors. So I cannot furthur track the cause. But Chrome devtools show: 
```
main.js?1676357479:2 WebSocket connection to 'wss://MyHost:6002/app/biigle?protocol=7&client=js&version=7.5.0&flash=false' failed: WebSocket is closed before the connection is established."
```
It seems all the modifications still been saved, but just the spinner hang-up the operations.  Any comment to debug/fix it will be appreciated. Thanks.

**Comment by mzur** [comment]
WebSocket support was introduced in [v3.50.0](https://github.com/biigle/core/releases/tag/v3.50.0). Did you update your configuration as described in the release notes?

If the configuration should be correct, do you have a firewall? It should allow requests to port `6002`.

You are correct that the Largo session is still saved. The WebSockets connection is only used to show the success message.

**Reply by cywhale** [reply]
I think my .env had been updated 
```
BROADCAST_DRIVER=pusher
PUSHER_HOST=websockets
PUSHER_PORT=6001
PUSHER_APP_ID=biigle
PUSHER_APP_KEY=biigle
PUSHER_APP_SECRET=xxxxxxxxxxxx

SOKETI_DEFAULT_APP_ID="${PUSHER_APP_ID}"
SOKETI_DEFAULT_APP_KEY="${PUSHER_APP_KEY}"
SOKETI_DEFAULT_APP_SECRET="${PUSHER_APP_SECRET}"

MIX_PUSHER_APP_KEY="${PUSHER_APP_KEY}"
MIX_PUSHER_APP_HOST=MyHost
MIX_PUSHER_APP_TLS=true
MIX_PUSHER_PORT=6002

# in docker-compose.yaml
  websockets:
    restart: always
    image: biigle/websockets-dist
    ports:
      - 6001:6001

  web:
#...
    ports:
      - 6002:6002

```
The websocket failed message is "WebSocket is closed before the connection is established."
I found many similar cases in Laravel on stackoverflow. Try to debug it furthur.

**Reply by mzur** [reply]
What about a firewall? Do you see any issues in the network tab of the Chrome devtools?

**Reply by cywhale** [reply]
> What about a firewall? Do you see any issues in the network tab of the Chrome devtools?

No firewall rule on this port (ufw is also not active)
```
$netstat -tulpn
tcp        0      0 0.0.0.0:6379            0.0.0.0:*               LISTEN      -                   
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      -                   
tcp        0      0 0.0.0.0:8081            0.0.0.0:*               LISTEN      -                   
tcp        0      0 0.0.0.0:6001            0.0.0.0:*               LISTEN      -                   
tcp        0      0 0.0.0.0:6002            0.0.0.0:*               LISTEN      -         
.... 
```

The message on Chrome devtool on Network tab is

![https://github.com/cywhale/ODB/blob/master/ROV/biigle/clip_202303_01.png](https://raw.githubusercontent.com/cywhale/ODB/master/ROV/biigle/clip_202303_01.png)

```
Headers: wss://seaimage.oc.ntu.edu.tw:6002/app/biigle?protocol=7&client=js&version=7.5.0&flash=false

Payload:
protocol: 7
client: js
version: 7.5.0
flash: false

message:
WebSocket is closed before the connection is established.

trace stack
createWebSocket @ https://MyHost/assets/scripts/main.js?1676357479:2
getSocket @ https://MyHost/assets/scripts/main.js?1676357479:2
e.connect @ https://MyHOst/assets/scripts/main.js?1676357479:2
....
```

Error on Console tab 
```
main.js?1676357479:2 WebSocket connection to 'wss://MyHost:6002/app/biigle?protocol=7&client=js&version=7.5.0&flash=false' failed: WebSocket is closed before the connection is established.
```

Have tried some websocket suggested configurations on Apache2/NGINX, but seems no effect for this.

**Reply by mzur** [reply]
Maybe it has something to do with the app being hosted from a subdirectory (`/app/biigle`)? Either the subdirectory should not be used in the websocket URL or the [Nginx configuration](https://github.com/biigle/core/blob/131612697d3c5922d1490b798bb36e2e18fee2f7/.docker/vhost.conf#L52-L82) must be changed to accomodate this?

**Reply by mzur** [reply, marked as answer]
fyi: The websocket configuration was simplified with the latest release. It now uses port 80 (or 443), same than the regular HTTP requests. No additional firewall rules are required any more. You shouldn't have to update your setup unless you made custom changes to the `vhost.conf` files.

**Reply by cywhale** [reply]
It works!! Thanks^^ Thanks for the kind reminder and this handy upgrade!!
